jax/lax/lax.py

```diff
@@ -573,6 +572,94 @@ def conv_general_dilated(
+  dimension is therefore the size of each patch, i.e.
+  `np.prod(filter_shape) * lhs.shape[lhs_spec.index('C')]`, where
+  `lhs_spec == dimension_numbers[0]`.
+
+  Docstring below adapted from `jax.lax.conv_general_dilated`.
+
+  See Also:
+    https://www.tensorflow.org/xla/operation_semantics#conv_convolution
+
+  Args:
+    lhs: a rank `n+2` dimensional input array.
+    filter_shape: a sequence of `n` integers, representing the receptive window
+      spatial shape in the order as specified in
+      `rhs_spec = dimension_numbers[1]`.
+    window_strides: a sequence of `n` integers, representing the inter-window
+      strides.
+    padding: either the string `'SAME'`, the string `'VALID'`, or a sequence of
+      `n` `(low, high)` integer pairs that give the padding to apply before and
+      after each spatial dimension.
+    lhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `lhs`. LHS dilation
+      is also known as transposed convolution.
+    rhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `rhs`. RHS dilation
+      is also known as atrous convolution.
+    dimension_numbers: either `None`, or a 3-tuple
+      `(lhs_spec, rhs_spec, out_spec)`, where each element is a string
+      of length `n+2`. `None` defaults to `("NCHWD..., OIHWD..., NCHWD...")`.
+    precision: Optional. Either ``None``, which means the default precision for
+      the backend, or a ``lax.Precision`` enum value (``Precision.DEFAULT``,
+      ``Precision.HIGH`` or ``Precision.HIGHEST``).
+
+  Returns:
+    A rank `n+2` array containing the flattened image patches in the output
+    channel (`"C"`) dimension. For example if
+    `dimension_numbers = ("NcHW", "OIwh", "CNHW")`, the output has dimension
+    numbers `"CNHW" = "{cwh}NHW"`, with the size of dimension `"C"` equal to
+    the size of each patch
+    (`np.prod(filter_shape) * lhs.shape[lhs_spec.index('C')]`).
+
+  """
+  filter_shape = tuple(filter_shape)
+  dimension_numbers = conv_dimension_numbers(lhs.shape, (1, 1) + filter_shape,
+                                             dimension_numbers)
+
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+
+  spatial_size = np.prod(filter_shape)
+  n_channels = lhs.shape[lhs_spec[1]]
+
+  # Move separate `lhs` spatial locations into separate `rhs` channels.
+  rhs = _eye(lhs.dtype, (spatial_size,) * 2, 0).reshape(filter_shape * 2)
+
+  rhs = rhs.reshape((spatial_size, 1) + filter_shape)
+  rhs = np.tile(rhs, (n_channels,) + (1,) * (rhs.ndim - 1))
+  rhs = np.moveaxis(rhs, (0, 1), (rhs_spec[0], rhs_spec[1]))
+
+  out = conv_general_dilated(
+      lhs=lhs,
+      rhs=rhs,
+      window_strides=window_strides,
+      padding=padding,
+      lhs_dilation=lhs_dilation,
+      rhs_dilation=rhs_dilation,
+      dimension_numbers=dimension_numbers,
+      precision=precision,
```

Comment:
This would be a nice place to use the newly support option to use tuple for precision:
https://github.com/google/jax/pull/4586

The LHS precision can always be `Precision.DEFAULT`, because 0 and 1 can be represented exactly in bfloat16. This should make `conv_general_dilated_patches` twice as fast on TPU when using full float32 precision.

Reply:
Done! (set `Precision.DEFAULT` for `rhs`, not `lhs`, right?)